File: backend/middleware/api_versioning.py
"""
API Versioning Middleware for SizeWise Suite Backend

Provides comprehensive API versioning with backward compatibility,
deprecation management, and migration support for Flask applications.
"""

import re
import logging
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Tuple, Any
from functools import wraps
from flask import request, jsonify, g, current_app
from dataclasses import dataclass

# =============================================================================
# Version Data Classes
# =============================================================================

@dataclass
class ApiVersion:
    major: int
    minor: int
    patch: int
    prerelease: Optional[str] = None
    build: Optional[str] = None
    
    def __str__(self) -> str:
        version_str = f"{self.major}.{self.minor}.{self.patch}"
        if self.prerelease:
            version_str += f"-{self.prerelease}"
        if self.build:
            version_str += f"+{self.build}"
        return version_str
    
    def __eq__(self, other) -> bool:
        if not isinstance(other, ApiVersion):
            return False
        return (self.major, self.minor, self.patch, self.prerelease) == \
               (other.major, other.minor, other.patch, other.prerelease)
    
    def __lt__(self, other) -> bool:
        if not isinstance(other, ApiVersion):
            return NotImplemented
        return (self.major, self.minor, self.patch) < (other.major, other.minor, other.patch)

@dataclass
class VersionedEndpoint:
    path: str
    version: ApiVersion
    method: str
    handler: callable
    deprecated: bool = False
    deprecation_date: Optional[datetime] = None
    removal_date: Optional[datetime] = None
    replaced_by: Optional[str] = None
    description: str = ""
    changes: List[str] = None
    
    def __post_init__(self):
        if self.changes is None:
            self.changes = []

@dataclass
class VersioningConfig:
    current_version: ApiVersion
    supported_versions: List[ApiVersion]
    default_version: ApiVersion
    versioning_strategy: str = 'url'  # 'url', 'header', 'query', 'content-type'
    strict_versioning: bool = False
    deprecation_warning_months: int = 3
    deprecation_support_months: int = 6

# =============================================================================
# Version Utilities
# =============================================================================

class VersionUtils:
    @staticmethod
    def parse_version(version_string: str) -> ApiVersion:
        """Parse version string to ApiVersion object."""
        pattern = r'^(\d+)\.(\d+)\.(\d+)(?:-([a-zA-Z0-9.-]+))?(?:\+([a-zA-Z0-9.-]+))?$'
        match = re.match(pattern, version_string)
        
        if not match:
            raise ValueError(f"Invalid version format: {version_string}")
        
        return ApiVersion(
            major=int(match.group(1)),
            minor=int(match.group(2)),
            patch=int(match.group(3)),
            prerelease=match.group(4),
            build=match.group(5)
        )
    
    @staticmethod
    def is_compatible(requested: ApiVersion, available: ApiVersion) -> bool:
        """Check if requested version is compatible with available version."""
        # Major version must match for compatibility
        if requested.major != available.major:
            return False
        
        # Minor version can be backward compatible
        if requested.minor > available.minor:
            return False
        
        return True
    
    @staticmethod
    def compare_versions(a: ApiVersion, b: ApiVersion) -> int:
        """Compare two versions. Returns -1, 0, or 1."""
        if a < b:
            return -1
        elif a == b:
            return 0
        else:
            return 1

# =============================================================================
# API Version Manager
# =============================================================================

class ApiVersionManager:
    def __init__(self, config: VersioningConfig):
        self.config = config
        self.endpoints: Dict[str, List[VersionedEndpoint]] = {}
        self.logger = logging.getLogger(__name__)
    
    def register_endpoint(self, endpoint: VersionedEndpoint) -> None:
        """Register a versioned endpoint."""
        key = f"{endpoint.method}:{endpoint.path}"
        
        if key not in self.endpoints:
            self.endpoints[key] = []
        
        # Check for duplicate versions
        for existing in self.endpoints[key]:
            if existing.version == endpoint.version:
                raise ValueError(f"Endpoint {key} already registered for version {endpoint.version}")
        
        self.endpoints[key].append(endpoint)
        # Sort by version (latest first)
        self.endpoints[key].sort(key=lambda e: (e.version.major, e.version.minor, e.version.patch), reverse=True)
    
    def get_endpoint(self, method: str, path: str, requested_version: Optional[ApiVersion] = None) -> Optional[VersionedEndpoint]:
        """Get endpoint for specific version."""
        key = f"{method}:{path}"
        endpoints = self.endpoints.get(key, [])
        
        if not endpoints:
            return None
        
        target_version = requested_version or self.config.default_version
        
        # Find exact match first
        for endpoint in endpoints:
            if endpoint.version == target_version:
                return endpoint
        
        # Find compatible version (backward compatibility)
        for endpoint in endpoints:
            if VersionUtils.is_compatible(target_version, endpoint.version):
                return endpoint
        
        # Return latest if no compatible version found and strict versioning is disabled
        if not self.config.strict_versioning:
            return endpoints[0]  # Latest version (sorted)
        
        return None
    
    def extract_version_from_request(self) -> Optional[ApiVersion]:
        """Extract version from current Flask request."""
        try:
            if self.config.versioning_strategy == 'url':
                return self._extract_version_from_url()
            elif self.config.versioning_strategy == 'header':
                return self._extract_version_from_header()
            elif self.config.versioning_strategy == 'query':
                return self._extract_version_from_query()
            elif self.config.versioning_strategy == 'content-type':
                return self._extract_version_from_content_type()
        except Exception as e:
            self.logger.warning(f"Failed to extract version from request: {e}")
        
        return None
    
    def generate_version_headers(self, endpoint: VersionedEndpoint) -> Dict[str, str]:
        """Generate version-related response headers."""
        headers = {
            'API-Version': str(endpoint.version),
            'API-Supported-Versions': ', '.join(str(v) for v in self.config.supported_versions)
        }
        
        if endpoint.deprecated:
            headers['API-Deprecation'] = 'true'
            
            if endpoint.deprecation_date:
                headers['API-Deprecation-Date'] = endpoint.deprecation_date.isoformat()
            
            if endpoint.removal_date:
                headers['API-Removal-Date'] = endpoint.removal_date.isoformat()
            
            if endpoint.replaced_by:
                headers['API-Replaced-By'] = endpoint.replaced_by
        
        return headers
    
    def _extract_version_from_url(self) -> Optional[ApiVersion]:
        """Extract version from URL path."""
        path = request.path
        match = re.search(r'/v(\d+)(?:\.(\d+))?(?:\.(\d+))?', path)
        if match:
            return ApiVersion(
                major=int(match.group(1)),
                minor=int(match.group(2) or '0'),
                patch=int(match.group(3) or '0')
            )
        return None
    
    def _extract_version_from_header(self) -> Optional[ApiVersion]:
        """Extract version from API-Version header."""
        version_header = request.headers.get('API-Version')
        if version_header:
            return VersionUtils.parse_version(version_header)
        return None
    
    def _extract_version_from_query(self) -> Optional[ApiVersion]:
        """Extract version from query parameter."""
        version_param = request.args.get('version') or request.args.get('v')
        if version_param:
            return VersionUtils.parse_version(version_param)
        return None
    
    def _extract_version_from_content_type(self) -> Optional[ApiVersion]:
        """Extract version from Content-Type header."""
        content_type = request.headers.get('Content-Type', '')
        match = re.search(r'application/vnd\.sizewise\.v(\d+)(?:\.(\d+))?(?:\.(\d+))?\+json', content_type)
        if match:
            return ApiVersion(
                major=int(match.group(1)),
                minor=int(match.group(2) or '0'),
                patch=int(match.group(3) or '0')
            )
        return None

# =============================================================================
# Flask Middleware and Decorators
# =============================================================================

# Global version manager instance
version_manager: Optional[ApiVersionManager] = None

def init_api_versioning(app, config: VersioningConfig):
    """Initialize API versioning for Flask app."""
    global version_manager
    version_manager = ApiVersionManager(config)
    
    @app.before_request
    def before_request():
        """Extract and validate API version before each request."""
        if not request.path.startswith('/api/'):
            return  # Skip non-API requests
        
        requested_version = version_manager.extract_version_from_request()
        g.api_version = requested_version or config.default_version
        g.version_manager = version_manager

def versioned_endpoint(version: str, path: str, methods: List[str] = None, 
                      deprecated: bool = False, deprecation_date: Optional[str] = None,
                      removal_date: Optional[str] = None, replaced_by: Optional[str] = None,
                      description: str = ""):
    """Decorator for versioned API endpoints."""
    if methods is None:
        methods = ['GET']
    
    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            # Get version manager from Flask g
            vm = getattr(g, 'version_manager', None)
            if not vm:
                return jsonify({'error': 'API versioning not initialized'}), 500
            
            # Get current request version
            current_version = getattr(g, 'api_version', None)
            if not current_version:
                return jsonify({'error': 'API version not specified'}), 400
            
            # Find appropriate endpoint
            endpoint = vm.get_endpoint(request.method, path, current_version)
            if not endpoint:
                return jsonify({
                    'error': 'API version not supported',
                    'requested_version': str(current_version),
                    'supported_versions': [str(v) for v in vm.config.supported_versions]
                }), 400
            
            # Execute the function
            response = func(*args, **kwargs)
            
            # Add version headers to response
            if hasattr(response, 'headers'):
                version_headers = vm.generate_version_headers(endpoint)
                for key, value in version_headers.items():
                    response.headers[key] = value
            
            return response
        
        # Register the endpoint
        if version_manager:
            parsed_version = VersionUtils.parse_version(version)
            dep_date = datetime.fromisoformat(deprecation_date) if deprecation_date else None
            rem_date = datetime.fromisoformat(removal_date) if removal_date else None
            
            for method in methods:
                endpoint = VersionedEndpoint(
                    path=path,
                    version=parsed_version,
                    method=method,
                    handler=func,
                    deprecated=deprecated,
                    deprecation_date=dep_date,
                    removal_date=rem_date,
                    replaced_by=replaced_by,
                    description=description
                )
                version_manager.register_endpoint(endpoint)
        
        return wrapper
    return decorator

# =============================================================================
# Default Configuration for SizeWise Suite
# =============================================================================

def get_default_versioning_config() -> VersioningConfig:
    """Get default versioning configuration for SizeWise Suite."""
    return VersioningConfig(
        current_version=VersionUtils.parse_version('1.0.0'),
        supported_versions=[
            VersionUtils.parse_version('1.0.0')
        ],
        default_version=VersionUtils.parse_version('1.0.0'),
        versioning_strategy='url',
        strict_versioning=False,
        deprecation_warning_months=3,
        deprecation_support_months=6
    )

# =============================================================================
# Utility Functions
# =============================================================================

def get_api_version_info() -> Dict[str, Any]:
    """Get API version information for the /api/info endpoint."""
    if not version_manager:
        return {}
    
    config = version_manager.config
    return {
        'current_version': str(config.current_version),
        'supported_versions': [str(v) for v in config.supported_versions],
        'default_version': str(config.default_version),
        'versioning_strategy': config.versioning_strategy,
        'strict_versioning': config.strict_versioning,
        'deprecation_policy': {
            'warning_period_months': config.deprecation_warning_months,
            'support_period_months': config.deprecation_support_months
        }
    }

File: shared/api/ApiVersioning.ts
/**
 * API Versioning System for SizeWise Suite
 * 
 * Provides comprehensive API versioning with backward compatibility,
 * deprecation management, and migration support.
 */

// =============================================================================
// Version Types and Interfaces
// =============================================================================

export interface ApiVersion {
  major: number;
  minor: number;
  patch: number;
  prerelease?: string;
  build?: string;
}

export interface VersionedEndpoint {
  path: string;
  version: ApiVersion;
  method: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH';
  deprecated?: boolean;
  deprecationDate?: Date;
  removalDate?: Date;
  replacedBy?: string;
  description: string;
  changes?: string[];
}

export interface ApiVersionConfig {
  currentVersion: ApiVersion;
  supportedVersions: ApiVersion[];
  deprecationPolicy: {
    warningPeriodMonths: number;
    supportPeriodMonths: number;
    notificationMethods: ('header' | 'response' | 'email')[];
  };
  versioningStrategy: 'url' | 'header' | 'query' | 'content-type';
  defaultVersion: ApiVersion;
  strictVersioning: boolean;
}

export interface VersionCompatibility {
  fromVersion: ApiVersion;
  toVersion: ApiVersion;
  compatible: boolean;
  breakingChanges: string[];
  migrationGuide?: string;
  autoMigration?: boolean;
}

// =============================================================================
// Version Utilities
// =============================================================================

export class VersionUtils {
  /**
   * Parse version string to ApiVersion object
   */
  static parseVersion(versionString: string): ApiVersion {
    const regex = /^(\d+)\.(\d+)\.(\d+)(?:-([a-zA-Z0-9.-]+))?(?:\+([a-zA-Z0-9.-]+))?$/;
    const match = versionString.match(regex);
    
    if (!match) {
      throw new Error(`Invalid version format: ${versionString}`);
    }
    
    return {
      major: parseInt(match[1], 10),
      minor: parseInt(match[2], 10),
      patch: parseInt(match[3], 10),
      prerelease: match[4],
      build: match[5]
    };
  }
  
  /**
   * Convert ApiVersion to string
   */
  static versionToString(version: ApiVersion): string {
    let versionString = `${version.major}.${version.minor}.${version.patch}`;
    
    if (version.prerelease) {
      versionString += `-${version.prerelease}`;
    }
    
    if (version.build) {
      versionString += `+${version.build}`;
    }
    
    return versionString;
  }
  
  /**
   * Compare two versions
   */
  static compareVersions(a: ApiVersion, b: ApiVersion): number {
    // Compare major
    if (a.major !== b.major) {
      return a.major - b.major;
    }
    
    // Compare minor
    if (a.minor !== b.minor) {
      return a.minor - b.minor;
    }
    
    // Compare patch
    if (a.patch !== b.patch) {
      return a.patch - b.patch;
    }
    
    // Compare prerelease
    if (a.prerelease && !b.prerelease) return -1;
    if (!a.prerelease && b.prerelease) return 1;
    if (a.prerelease && b.prerelease) {
      return a.prerelease.localeCompare(b.prerelease);
    }
    
    return 0;
  }
  
  /**
   * Check if version is compatible
   */
  static isCompatible(requestedVersion: ApiVersion, availableVersion: ApiVersion): boolean {
    // Major version must match for compatibility
    if (requestedVersion.major !== availableVersion.major) {
      return false;
    }
    
    // Minor version can be backward compatible
    if (requestedVersion.minor > availableVersion.minor) {
      return false;
    }
    
    return true;
  }
  
  /**
   * Get latest version from array
   */
  static getLatestVersion(versions: ApiVersion[]): ApiVersion {
    if (versions.length === 0) {
      throw new Error('No versions provided');
    }
    
    return versions.reduce((latest, current) => 
      this.compareVersions(current, latest) > 0 ? current : latest
    );
  }
}

// =============================================================================
// API Version Manager
// =============================================================================

export class ApiVersionManager {
  private config: ApiVersionConfig;
  private endpoints: Map<string, VersionedEndpoint[]> = new Map();
  private compatibilityMatrix: VersionCompatibility[] = [];
  
  constructor(config: ApiVersionConfig) {
    this.config = config;
    this.initializeCompatibilityMatrix();
  }
  
  /**
   * Register a versioned endpoint
   */
  registerEndpoint(endpoint: VersionedEndpoint): void {
    const key = `${endpoint.method}:${endpoint.path}`;
    const existing = this.endpoints.get(key) || [];
    
    // Check for duplicate versions
    const duplicateVersion = existing.find(e => 
      VersionUtils.compareVersions(e.version, endpoint.version) === 0
    );
    
    if (duplicateVersion) {
      throw new Error(`Endpoint ${key} already registered for version ${VersionUtils.versionToString(endpoint.version)}`);
    }
    
    existing.push(endpoint);
    existing.sort((a, b) => VersionUtils.compareVersions(b.version, a.version)); // Latest first
    this.endpoints.set(key, existing);
  }
  
  /**
   * Get endpoint for specific version
   */
  getEndpoint(method: string, path: string, requestedVersion?: ApiVersion): VersionedEndpoint | null {
    const key = `${method}:${path}`;
    const endpoints = this.endpoints.get(key);
    
    if (!endpoints || endpoints.length === 0) {
      return null;
    }
    
    const targetVersion = requestedVersion || this.config.defaultVersion;
    
    // Find exact match first
    const exactMatch = endpoints.find(e => 
      VersionUtils.compareVersions(e.version, targetVersion) === 0
    );
    
    if (exactMatch) {
      return exactMatch;
    }
    
    // Find compatible version (backward compatibility)
    const compatibleEndpoint = endpoints.find(e => 
      VersionUtils.isCompatible(targetVersion, e.version)
    );
    
    if (compatibleEndpoint) {
      return compatibleEndpoint;
    }
    
    // Return latest if no compatible version found and strict versioning is disabled
    if (!this.config.strictVersioning) {
      return endpoints[0]; // Latest version (sorted)
    }
    
    return null;
  }
  
  /**
   * Get all supported versions for an endpoint
   */
  getSupportedVersions(method: string, path: string): ApiVersion[] {
    const key = `${method}:${path}`;
    const endpoints = this.endpoints.get(key) || [];
    return endpoints.map(e => e.version);
  }
  
  /**
   * Check if endpoint is deprecated
   */
  isDeprecated(method: string, path: string, version: ApiVersion): boolean {
    const endpoint = this.getEndpoint(method, path, version);
    return endpoint?.deprecated || false;
  }
  
  /**
   * Get deprecation info
   */
  getDeprecationInfo(method: string, path: string, version: ApiVersion): {
    deprecated: boolean;
    deprecationDate?: Date;
    removalDate?: Date;
    replacedBy?: string;
    warningMessage?: string;
  } {
    const endpoint = this.getEndpoint(method, path, version);
    
    if (!endpoint || !endpoint.deprecated) {
      return { deprecated: false };
    }
    
    const warningMessage = this.generateDeprecationWarning(endpoint);
    
    return {
      deprecated: true,
      deprecationDate: endpoint.deprecationDate,
      removalDate: endpoint.removalDate,
      replacedBy: endpoint.replacedBy,
      warningMessage
    };
  }
  
  /**
   * Extract version from request
   */
  extractVersionFromRequest(request: {
    url?: string;
    headers?: Record<string, string>;
    query?: Record<string, string>;
  }): ApiVersion | null {
    switch (this.config.versioningStrategy) {
      case 'url':
        return this.extractVersionFromUrl(request.url || '');
      case 'header':
        return this.extractVersionFromHeader(request.headers || {});
      case 'query':
        return this.extractVersionFromQuery(request.query || {});
      case 'content-type':
        return this.extractVersionFromContentType(request.headers || {});
      default:
        return null;
    }
  }
  
  /**
   * Generate version response headers
   */
  generateVersionHeaders(endpoint: VersionedEndpoint): Record<string, string> {
    const headers: Record<string, string> = {
      'API-Version': VersionUtils.versionToString(endpoint.version),
      'API-Supported-Versions': this.config.supportedVersions
        .map(v => VersionUtils.versionToString(v))
        .join(', ')
    };
    
    if (endpoint.deprecated) {
      headers['API-Deprecation'] = 'true';
      
      if (endpoint.deprecationDate) {
        headers['API-Deprecation-Date'] = endpoint.deprecationDate.toISOString();
      }
      
      if (endpoint.removalDate) {
        headers['API-Removal-Date'] = endpoint.removalDate.toISOString();
      }
      
      if (endpoint.replacedBy) {
        headers['API-Replaced-By'] = endpoint.replacedBy;
      }
    }
    
    return headers;
  }
  
  // =============================================================================
  // Private Methods
  // =============================================================================
  
  private initializeCompatibilityMatrix(): void {
    // Initialize compatibility matrix for SizeWise Suite versions
    this.compatibilityMatrix = [
      {
        fromVersion: VersionUtils.parseVersion('1.0.0'),
        toVersion: VersionUtils.parseVersion('1.1.0'),
        compatible: true,
        breakingChanges: [],
        autoMigration: true
      },
      {
        fromVersion: VersionUtils.parseVersion('1.1.0'),
        toVersion: VersionUtils.parseVersion('2.0.0'),
        compatible: false,
        breakingChanges: [
          'Authentication endpoint moved from /auth to /api/auth',
          'Response format changed for calculation endpoints',
          'New required fields in project creation'
        ],
        migrationGuide: '/docs/migration/v1-to-v2'
      }
    ];
  }
  
  private extractVersionFromUrl(url: string): ApiVersion | null {
    const versionMatch = url.match(/\/v(\d+)(?:\.(\d+))?(?:\.(\d+))?/);
    if (versionMatch) {
      return {
        major: parseInt(versionMatch[1], 10),
        minor: parseInt(versionMatch[2] || '0', 10),
        patch: parseInt(versionMatch[3] || '0', 10)
      };
    }
    return null;
  }
  
  private extractVersionFromHeader(headers: Record<string, string>): ApiVersion | null {
    const versionHeader = headers['API-Version'] || headers['api-version'];
    if (versionHeader) {
      try {
        return VersionUtils.parseVersion(versionHeader);
      } catch {
        return null;
      }
    }
    return null;
  }
  
  private extractVersionFromQuery(query: Record<string, string>): ApiVersion | null {
    const version = query.version || query.v;
    if (version) {
      try {
        return VersionUtils.parseVersion(version);
      } catch {
        return null;
      }
    }
    return null;
  }
  
  private extractVersionFromContentType(headers: Record<string, string>): ApiVersion | null {
    const contentType = headers['Content-Type'] || headers['content-type'];
    if (contentType) {
      const versionMatch = contentType.match(/application\/vnd\.sizewise\.v(\d+)(?:\.(\d+))?(?:\.(\d+))?\+json/);
      if (versionMatch) {
        return {
          major: parseInt(versionMatch[1], 10),
          minor: parseInt(versionMatch[2] || '0', 10),
          patch: parseInt(versionMatch[3] || '0', 10)
        };
      }
    }
    return null;
  }
  
  private generateDeprecationWarning(endpoint: VersionedEndpoint): string {
    let warning = `API endpoint ${endpoint.method} ${endpoint.path} version ${VersionUtils.versionToString(endpoint.version)} is deprecated.`;
    
    if (endpoint.removalDate) {
      warning += ` It will be removed on ${endpoint.removalDate.toDateString()}.`;
    }
    
    if (endpoint.replacedBy) {
      warning += ` Please use ${endpoint.replacedBy} instead.`;
    }
    
    return warning;
  }
}

// =============================================================================
// Migration System
// =============================================================================

export interface ApiMigration {
  fromVersion: ApiVersion;
  toVersion: ApiVersion;
  description: string;
  breakingChanges: string[];
  migrationSteps: MigrationStep[];
  autoMigration: boolean;
  migrationGuideUrl?: string;
}

export interface MigrationStep {
  step: number;
  title: string;
  description: string;
  codeExample?: string;
  required: boolean;
  estimatedTime?: string;
}

export class ApiMigrationManager {
  private migrations: Map<string, ApiMigration> = new Map();

  /**
   * Register a migration path
   */
  registerMigration(migration: ApiMigration): void {
    const key = `${VersionUtils.versionToString(migration.fromVersion)}->${VersionUtils.versionToString(migration.toVersion)}`;
    this.migrations.set(key, migration);
  }

  /**
   * Get migration for version upgrade
   */
  getMigration(fromVersion: ApiVersion, toVersion: ApiVersion): ApiMigration | null {
    const key = `${VersionUtils.versionToString(fromVersion)}->${VersionUtils.versionToString(toVersion)}`;
    return this.migrations.get(key) || null;
  }

  /**
   * Get all available migrations
   */
  getAllMigrations(): ApiMigration[] {
    return Array.from(this.migrations.values());
  }

  /**
   * Check if migration is available
   */
  hasMigration(fromVersion: ApiVersion, toVersion: ApiVersion): boolean {
    return this.getMigration(fromVersion, toVersion) !== null;
  }

  /**
   * Get migration path (multi-step migrations)
   */
  getMigrationPath(fromVersion: ApiVersion, toVersion: ApiVersion): ApiMigration[] {
    const path: ApiMigration[] = [];
    let currentVersion = fromVersion;

    while (VersionUtils.compareVersions(currentVersion, toVersion) < 0) {
      let nextMigration: ApiMigration | null = null;

      // Find the next migration step
      for (const migration of this.migrations.values()) {
        if (VersionUtils.compareVersions(migration.fromVersion, currentVersion) === 0 &&
            VersionUtils.compareVersions(migration.toVersion, toVersion) <= 0) {
          if (!nextMigration ||
              VersionUtils.compareVersions(migration.toVersion, nextMigration.toVersion) < 0) {
            nextMigration = migration;
          }
        }
      }

      if (!nextMigration) {
        break; // No migration path found
      }

      path.push(nextMigration);
      currentVersion = nextMigration.toVersion;
    }

    return path;
  }
}

// =============================================================================
// Default Configuration
// =============================================================================

export const defaultApiVersionConfig: ApiVersionConfig = {
  currentVersion: VersionUtils.parseVersion('1.0.0'),
  supportedVersions: [
    VersionUtils.parseVersion('1.0.0')
  ],
  deprecationPolicy: {
    warningPeriodMonths: 3,
    supportPeriodMonths: 6,
    notificationMethods: ['header', 'response']
  },
  versioningStrategy: 'url',
  defaultVersion: VersionUtils.parseVersion('1.0.0'),
  strictVersioning: false
};

// =============================================================================
// SizeWise Suite Migration Definitions
// =============================================================================

export const sizeWiseMigrations: ApiMigration[] = [
  {
    fromVersion: VersionUtils.parseVersion('1.0.0'),
    toVersion: VersionUtils.parseVersion('1.1.0'),
    description: 'Minor update with enhanced HVAC calculations and new validation endpoints',
    breakingChanges: [],
    autoMigration: true,
    migrationSteps: [
      {
        step: 1,
        title: 'Update API calls to use new validation endpoints',
        description: 'New validation endpoints provide enhanced error reporting',
        codeExample: `
// Before
POST /api/calculations/validate

// After (optional, backward compatible)
POST /api/validation/hvac-parameters
        `,
        required: false,
        estimatedTime: '5 minutes'
      }
    ]
  },
  {
    fromVersion: VersionUtils.parseVersion('1.1.0'),
    toVersion: VersionUtils.parseVersion('2.0.0'),
    description: 'Major update with restructured authentication and enhanced project management',
    breakingChanges: [
      'Authentication endpoints moved from /auth to /api/auth',
      'Project creation requires additional metadata fields',
      'Response format changed for calculation endpoints',
      'New required authentication headers'
    ],
    autoMigration: false,
    migrationGuideUrl: '/docs/migration/v1-to-v2',
    migrationSteps: [
      {
        step: 1,
        title: 'Update authentication endpoints',
        description: 'All authentication endpoints have been moved to /api/auth prefix',
        codeExample: `
// Before
POST /auth/login
POST /auth/register

// After
POST /api/auth/login
POST /api/auth/register
        `,
        required: true,
        estimatedTime: '15 minutes'
      },
      {
        step: 2,
        title: 'Update project creation calls',
        description: 'Project creation now requires additional metadata',
        codeExample: `
// Before
{
  "name": "Project Name",
  "description": "Description"
}

// After
{
  "name": "Project Name",
  "description": "Description",
  "metadata": {
    "created_by": "user_id",
    "project_type": "hvac_design",
    "version": "2.0.0"
  }
}
        `,
        required: true,
        estimatedTime: '30 minutes'
      },
      {
        step: 3,
        title: 'Update calculation response handling',
        description: 'Calculation responses now include additional validation data',
        codeExample: `
// Before
{
  "result": { ... },
  "success": true
}

// After
{
  "result": { ... },
  "success": true,
  "validation": {
    "warnings": [...],
    "compliance": {...}
  },
  "metadata": {
    "calculation_time": "2024-01-01T00:00:00Z",
    "version": "2.0.0"
  }
}
        `,
        required: true,
        estimatedTime: '20 minutes'
      }
    ]
  }
];

// Initialize migration manager with default migrations
export const defaultMigrationManager = new ApiMigrationManager();
sizeWiseMigrations.forEach(migration => {
  defaultMigrationManager.registerMigration(migration);
});
